sdk/identity/identity/src/msal/msalClient.ts

```diff
@@ -0,0 +1,260 @@
+    scopes: string[],
+    options?: GetTokenOptions,
+  ): Promise<msal.AuthenticationResult> {
+    if (state.cachedAccount === null) {
+      const cache = app.getTokenCache();
+      const accounts = await cache.getAllAccounts();
+
+      if (accounts === undefined || accounts.length === 0) {
+        throw new AuthenticationRequiredError({ scopes });
+      }
+
+      if (accounts.length > 1) {
+        msalLogger.info(`More than one account was found authenticated for this Client ID and Tenant ID.
+However, no "authenticationRecord" has been provided for this credential,
+therefore we're unable to pick between these accounts.
+A new login attempt will be requested, to ensure the correct account is picked.
+To work with multiple accounts for the same Client ID and Tenant ID, please provide an "authenticationRecord" when initializing a credential to prevent this from happening.`);
+        throw new AuthenticationRequiredError({ scopes });
+      }
+
+      state.cachedAccount = accounts[0];
+    }
+
+    // TODO: broker
+    msalLogger.getToken.info("Attempting to acquire token silently");
+    return app.acquireTokenSilent({
+      account: state.cachedAccount,
+      scopes,
+      claims: options?.claims,
+    });
+  }
+
+  /**
+   * Performs silent authentication using MSAL to acquire an access token.
+   * If silent authentication fails, falls back to interactive authentication.
+   *
+   * @param msalApp - The MSAL application instance.
+   * @param scopes - The scopes for which to acquire the access token.
+   * @param options - The options for acquiring the access token.
+   * @param onAuthenticationRequired - A callback function to handle interactive authentication when silent authentication fails.
+   * @returns A promise that resolves to an AccessToken object containing the access token and its expiration timestamp.
+   */
+  async function withSilentAuthentication(
+    msalApp: msal.ConfidentialClientApplication | msal.PublicClientApplication,
+    scopes: Array<string>,
+    options: GetTokenOptions,
+    onAuthenticationRequired: () => Promise<msal.AuthenticationResult | null>,
+  ): Promise<AccessToken> {
+    let response: msal.AuthenticationResult | null = null;
+    try {
+      response = await getTokenSilent(msalApp, scopes, options);
+    } catch (e: any) {
+      if (e.name !== "AuthenticationRequiredError") {
+        throw e;
+      }
+      if (createMsalClientOptions.disableAutomaticAuthentication) {
+        throw new AuthenticationRequiredError({
+          scopes,
+          getTokenOptions: options,
+          message:
+            "Automatic authentication has been disabled. You may call the authentication() method.",
+        });
+      }
+    }
+
+    // Silent authentication failed
+    if (response === null) {
+      response = await onAuthenticationRequired();
+    }
+
+    // At this point we should have a token, process it
+    ensureValidMsalToken(scopes, response, options);
```

Comment:
By the time we got here either the silent flow or the interactive flow completed , but the result processing is the same